staff/lab/timecard

```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env python3
+import argparse
+import collections
+import datetime
+
+import matplotlib.pyplot as plt
+from ocflib.lab import stats
+
+
+def generate_image(usage, users, dest, start, end):
+    def adjusted_hour(hour):
+        """Converts a 0-23 hour format into a 1-12 hour format"""
+        if hour == 0 or hour == 12:
+            return 12
+        else:
+            return hour % 12
+
+    def pm_or_am(hour):
+        return 'am' if hour <= 11 else 'pm'
+
+    hours = ['{}{}'.format(adjusted_hour(hour), pm_or_am(hour)) for hour in range(24)]
+    sums = [sum(usage[h]) for h in range(24)]
+
+    fig = plt.figure()
+    fig.set_size_inches(12, 4)
+
+    plt.bar(range(24), sums, color='b')
+    plt.xlim(0, 24)
+    plt.xticks(range(24), hours, fontsize=8)
+    plt.xlabel('Hour')
+    plt.ylabel('# Times Present')
+    plt.ylim(0, max(sums) * 1.15)
+
+    plt.title('OCF Lab Timecard for {} ({}-{})'.format(
+        ', '.join(users),
+        start.strftime('%Y-%m-%d'),
+        end.strftime('%Y-%m-%d')
+    ))
+    plt.savefig(dest, bbox_inches='tight')
+
+
+def get_usage_count(*users, start=None, end=None):
+    sem_start, sem_end = stats.semester_dates()
+    if start is None:
+        start = datetime.datetime.combine(sem_start, datetime.time())
+    if end is None:
+        end = datetime.datetime.combine(sem_end, datetime.time())
+
+    step = datetime.timedelta(minutes=60)
+    hours = collections.defaultdict(lambda: [0 for _ in range(7)])
+
+    cnx = stats.get_connection()
+    cursor = cnx.cursor()
+
+    query = """
+        SELECT `start`, `end`
+        FROM `staff_session_duration_public`
+        WHERE `user` IN %s
+            AND `start` >= %s
+            AND `end` <= %s
+            AND `duration` IS NOT NULL
+        ORDER BY start ASC"""
+    cursor.execute(query, (users, start, end))
+    logins = cursor.fetchall()
+
+    was_logged_in = dict()
+    for session in logins:
```

Comment:
It should be possible to combine this loop with the one below, no?

Reply:
The reason I did this loop before the actual counting one is to account for concurrent logins. By first setting a boolean value for every hour a user is logged in from the it easily accounts for a user being logged in for a given hour twice, whereas I don't think it'd be as intuitive to do so with sessions. We'd need to somehow keep track of previous sessions and check if they overlap. It could be done though.

Reply:
Isn't there an algorithm for combining overlapping intervals? Like, per-user, sort by start time, increment a counter when a session starts, and decrement when it ends?